tower-balance/src/lib.rs

```diff
@@ -26,166 +53,158 @@ pub enum Error<T, U> {
-                Err(_) => {}
+                Remove(key) => {
+                    self.endpoints.remove(&key);
+                    self.new_endpoints.remove(&key);
+                }
             }
-
-            self.inc_pos();
         }
 
-        // Not ready
-        Ok(Async::NotReady)
+        if self.endpoints.is_empty() {
+            Ok(Async::NotReady)
+        } else {
+            Ok(Async::Ready(()))
+        }
     }
+}
 
-    /// Returns a reference to the service to dispatch the next request to.
-    ///
-    /// It is expected that once `poll_ready` returns `Ready`, this function
-    /// must succeed.
-    ///
-    /// # Panics
+impl<D> Balance<
+    D,
+    load::WithConstant<D::Service>,
+    select::RoundRobin<D::Key, load::Constant<D::Service>>,
+>
+where
+    D: Discover,
+    D::Key: Clone,
+{
+    /// Creates a new balancer.
     ///
-    /// This function may panic if `poll_ready` did not return `Ready`
-    /// immediately prior.
-    fn next(&mut self) -> &mut T::Service {
-        let pos = self.pos;
-        self.inc_pos();
-
-        match self.endpoints.get_index_mut(pos) {
-            Some((_, val)) => val,
-            _ => panic!(),
-        }
+    /// The balancer selects each node in order. This order is not strictly enforced.
+    pub fn round_robin(discover: D) -> Self {
+        let load = load::WithConstant::min();
+        let select = select::RoundRobin::default();
+        Self::new(discover, load, select)
     }
+}
 
-    // ===== internal =====
+impl<D, R> Balance<
+    D,
+    load::WithPendingRequests<D::Service>,
+    select::PowerOfTwoChoices<D::Key, load::PendingRequests<D::Service>, R>,
+>
+where
+    D: Discover,
+    D::Key: Clone,
+    R: Rng,
+{
+    /// Creates a new balancer.
+    ///
+    /// The balancer selects the least-loaded endpoint using Mitzenmacher's _Power of Two
+    /// Choices_.
+    pub fn pending_requests_p2c(discover: D, rng: R) -> Self {
+        let load = load::WithPendingRequests::default();
+        let select = select::PowerOfTwoChoices::new(rng);
+        Self::new(discover, load, select)
+    }
+}
 
-    fn update_endpoints(&mut self) -> Result<(), T::DiscoverError> {
-        use tower_discover::Change::*;
+impl<D, L, S> Service for Balance<D, L, S>
+where
+    D: Discover,
+    D::Key: Clone,
+    L: WithLoad<Service = D::Service>,
+    S: Select<Key = D::Key, Service = L::PollLoad>,
+{
+    type Request = <L::PollLoad as Service>::Request;
+    type Response = <L::PollLoad as Service>::Response;
+    type Error = Error<<L::PollLoad as Service>::Error, D::DiscoverError>;
+    type Future = ResponseFuture<<L::PollLoad as Service>::Future, D::DiscoverError>;
 
-        loop {
-            let change = match try!(self.discover.poll()) {
-                Async::Ready(change) => change,
-                Async::NotReady => return Ok(()),
-            };
+    fn poll_ready(&mut self) -> Poll<(), Self::Error> {
+        try_ready!(self.poll_discover());
```

Comment:
we need to clear `ready_endpoint` before polling discover (in case there's a ready endpoint that is removed).